### PR TITLE
Add proper gif support (next try)

### PR DESCRIPTION
**Pull request by ace20022**

### CODE REVIEW

File: xbmc/pictures/Gif.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "guilib/iimage.h"
+#include "DllLibGif.h"
+#include <vector>
```
Comment (mkortstiege) [inline xbmc/pictures/Gif.h]: You need `#include <memory>` for `std::shared_ptr` (at least on Linux).

File: xbmc/pictures/Gif.cpp [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+    return false;
+
+  if (m_isAnimated < 0)
+  {
+    m_filename = file;
+    m_isAnimated = 0;
+
+    GifFileType* gif = NULL;
+    XFILE::CFile gifFile;
+    int err = 0;
```
Comment (mkortstiege) [inline xbmc/pictures/Gif.cpp]: unused variable

File: xbmc/pictures/Gif.cpp [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+  std::string message;
+  if (error)
+  {
+    message = StringUtils::Format(messageTemplate.append(" - %s").c_str(), error);
+  }
+  else
+  {
+    message = messageTemplate.append(" (reason unknown)");
+  }
+  CLog::Log(LOGERROR, message.c_str());
```
Comment (mkortstiege) [inline xbmc/pictures/Gif.cpp]: `%s` missing

**Comment by mkortstiege** [comment]
After fixing the minors it will compile on Linux. Will runtime test it later today.

File: xbmc/pictures/Gif.cpp (hunk 3 above)
Comment (ace20022) [inline xbmc/pictures/Gif.cpp]: ?

Comment (mkortstiege) [inline xbmc/pictures/Gif.cpp]: You're passing a string argument without having format specifier as target.

Comment (Paxxi) [inline xbmc/pictures/Gif.cpp]: he's doing that above so it's already preformatted, no need for the "%s"

File: xbmc/guilib/TextureManager.cpp [hunk 4]
```diff
@@ -349,42 +351,34 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
...
-        palette[i].x = 0xff;
-      // and set the transparent colour
-      if (AnimatedGifSet.m_vecimg[0]->Transparency && AnimatedGifSet.m_vecimg[0]->Transparent >= 0)
-        palette[AnimatedGifSet.m_vecimg[0]->Transparent].x = 0;
+      pMap = new CTextureMap(strTextureName, gif.Width(), gif.Height(), gif.GetNumLoops());
 
-      pMap = new CTextureMap(strTextureName, iWidth, iHeight, AnimatedGifSet.nLoops);
-
-      for (int iImage = 0; iImage < iImages; iImage++)
+      for (std::vector<Gif::FramePtr>::const_iterator frame = gif.GetFrames().begin(); frame != gif.GetFrames().end(); ++frame)
```
Comment (Paxxi) [inline xbmc/guilib/TextureManager.cpp]: use auto or the new foreach loop to make it easier to read

File: xbmc/pictures/Gif.cpp (hunk 3 above)
Comment (mkortstiege) [inline xbmc/pictures/Gif.cpp]: I get the following compile warning:
> Gif.cpp:574:31: warning: format not a string literal and no format arguments [-Wformat-security]

File: xbmc/pictures/Gif.cpp [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+    readBytes = 0;
+
+  if (len > buffSizeLeft)
+    readBytes = buffSizeLeft;
+
+  if (readBytes > 0)
+  {
+    unsigned char* src = ((Gifreader*)gif->UserData)->buffer + alreadyRead;
+    memcpy(gifbyte, src, readBytes);
+    ((Gifreader*)gif->UserData)->readPosition += readBytes;
```
Comment (Paxxi) [inline xbmc/pictures/Gif.cpp]: use c++ style cast

File: xbmc/pictures/Gif.cpp (hunk 3 above)
Comment (Paxxi) [inline xbmc/pictures/Gif.cpp]: ahh ok, well it's not an issue really but if it gives a warning we might as well try and change it, we don't really need more of those :)

**Comment by Paxxi** [comment]
In general it looks good @ace20022. The nitpicking about c++ style casts goes for everything.
Remember that .size() and sizeof returns uint64_t on win64 so might be better to use size_t for the variables instead of unsigned int.

I prefer using uint8_t vs unsigned char but that's just cosmetics and nothing stopping a merge :)

**Comment by MediaBrasil** [comment]
In 0_package_list, I think need replace line 17 with
giflib-5.1.0-win32.7z

**Comment by ace20022** [comment]
Just for the record I try to merge stuff from 4 or 5 different prs and sources into here. It's almost a bottomless pit. Currently only some textures work, e.g., the background of confluence. Todos are:
better animation timing, make use of the large image loader, caching, loading frames on demand (I had a 5mb gif that used 1,5gb and kodi crashed while copying the frames in the texture array), animation in the picture section. 
I doubt that I can do this all at once in this pr. I rather guess that I my interest will focus on the other topics I'm working on (bd-j, ext. audio, maybe dash) in a round robin way.

So the question is if this is wanted and can be merged before all the mentioned goals are reached. In the current state gif playback is broken, just set big_842bd7b827bb012efdfe9f0f52b00bddb9964bd0 as background in confluence.

**Comment by ace20022** [comment]
jenkins build this please

**Comment by Montellese** [comment]
@ace20022: jenkins is having issues so will have to wait with test builds

**Comment by Paxxi** [comment]
How broken is it, does it work in say pictures view?

In my opinion if it's no worse or even better than current support and doesn't cause crashes it's better to merge and then improve rather than having it linger as a forgotten PR

**Comment by Paxxi** [comment]
On a random side note your comment about using 1.5gb of memory got me thinking about this one https://www.bamsoftware.com/hacks/deflate.html . Truly evil :)

**Comment by ace20022** [comment]
> How broken is it, does it work in say pictures view?

Only the first frame gets displayed with and without this pr. Skin texture are fixed, I Don't have a screen rec software that works with dx11, does anyone know one?

**Comment by MartijnKaijser** [comment]
@ace20022 OBS works if you record the monitor. It can't record the Kodi window itself.

**Comment by ace20022** [comment]
@MartijnKaijser thx for the hint!

before: https://www.dropbox.com/s/tw35ev2h58xf7fu/counter_before.mkv?dl=0
after: https://www.dropbox.com/s/xal8eut0wzty8y5/counter_after.mkv?dl=0

before: https://www.dropbox.com/s/0nz11mvvtxkddly/before_clock.mkv?dl=0
after: https://www.dropbox.com/s/spa4y3v4kbhug53/after_clock.mkv?dl=0

https://github.com/ace20022/xbmc/commit/a9cf0a93a6051bcd0ff4dd84baa57c617cf91a77 might break bundled gif animations. Does anyone has an example of those?

**Comment by ace20022** [comment]
I know it won't build, but let's see why:

jenkins build this please

**Comment by ace20022** [comment]
@Memphiz there's an error with ios that I don't understand : http://jenkins.kodi.tv/job/IOS-ATV2/5150/consoleFull
Could you please have a look? IIRC the error started already with the first commit https://github.com/ace20022/xbmc/commit/f782a43fafa137b5b01407f1ea69ddc210eb9bab

**Comment by Memphiz** [comment]
drop that config site change https://github.com/ace20022/xbmc/commit/09327888d8dff454db0882dd0314eb948f2a0103#diff-2066a9e1e0661b86c32a7a5b3aff2e6f

**Comment by ace20022** [comment]
@Memphiz thx!
jenkins build this please

Will test giflib 4 and 5 with linux tomorrow and would like to merge it afterwards.
Would be nice if someone could test osx, rpi and android.

File: configure.ac [hunk 6]
```diff
@@ -1478,6 +1486,29 @@ else
...
+      AC_MSG_ERROR($libgif_not_found)
+      USE_LIBGIF=0
+    elif test "$use_libgif" != "no"; then
+      AC_MSG_NOTICE($libgif_not_found)
+      use_libgif="no"
+      USE_LIBGIF=0
+    fi
+   ])
+  if test "$use_libgif" != "no"; then
+    XB_FIND_SONAME([GIF], [gif], [use_libgif])
```
Comment (stefansaraev) [inline configure.ac]: as this is not dlopened at runtime, there is absolutely no reason to require distros having shared libgif.

**Comment by ace20022** [comment]
@wsnipex ping, good to merge?
I've tested kodi with giflib 4.1.6, some of my test files don't work and since I've implemented a rather strict error management nothing gets displayed for these.
If no objections, I'll merge the pr tomorrow afternoon.

jenkins build this please

**Comment by wsnipex** [comment]
what's the minimum required giflib version? If v4 doesn't work, we need to check the version in configure, which is a pain since there is no pkg-config file. See Texturepacker configure on how to do this.
Is there a reason for dyloading or could we also just link it? (could be done later to not hold this up)

**Comment by ace20022** [comment]
You insisted on version 4.1.6. There's no reason for dyloading.

**Comment by wsnipex** [comment]
ok then.

**Comment by HitcherUK** [comment]
@ace20022 This seems to have broken gifs that previously worked.

https://github.com/HitcherUK/skin.ftv/blob/master/media/views/equaliser.gif

**Comment by ace20022** [comment]
@HitcherUK that gif is a bit problematic. From http://giflib.sourceforge.net/whatsinagif/bits_and_bytes.html : 
>The next byte gives us the background color index. This byte is only meaningful if the global color table flag is 1, and if there is no global color table, this byte should be 0.. To understand it you have to remember the original "picture wall" rendering model for GIFs in which sub-images are composited onto a larger canvas. It represents which index in the global color table should be used for pixels on the virtual canvas that aren't overlayed by an image. GIFLIB supports reading and setting this byte, but modern viewers and browsers generally have no use for it.

ffplay and the old ms gif animator shows the same behaviour as kodi. But I'll have a closer look.

**Comment by HitcherUK** [comment]
Took me ages to build and get working in PS and I followed the guidelines in the wiki.

http://kodi.wiki/view/Texture_Attributes#Formats

**Pull request merged**